test/evp_extra_test2.c

```diff
@@ -347,6 +347,287 @@ static int test_d2i_PrivateKey_ex(void) {
+          && TEST_int_eq(EVP_PKEY_fromdata(gen_ctx, &pkey,
+                                           EVP_PKEY_KEYPAIR, key_params), 1)
+          && TEST_ptr(check_ctx = EVP_PKEY_CTX_new_from_pkey(mainctx, pkey,
+                                                        NULL))
+          && TEST_int_eq(EVP_PKEY_param_check(check_ctx), expected);
+    EVP_PKEY_CTX_free(check_ctx);
+    EVP_PKEY_CTX_free(gen_ctx);
+    EVP_PKEY_free(pkey);
+    return ret;
+}
+
+static int do_check_bn(OSSL_PARAM params[], const char *key,
+                       const unsigned char *expected, size_t expected_len)
+{
+    OSSL_PARAM *p;
+    BIGNUM *bn = NULL;
+    unsigned char buffer[256 + 1];
+    int ret, len;
+
+    ret = TEST_ptr(p = OSSL_PARAM_locate(params, key))
+          && TEST_true(OSSL_PARAM_get_BN(p, &bn))
+          && TEST_int_gt(len = BN_bn2binpad(bn, buffer, expected_len), 0)
+          && TEST_mem_eq(expected, expected_len, buffer, len);
+    BN_free(bn);
+    return ret;
+}
+
+static int do_check_int(OSSL_PARAM params[], const char *key, int expected)
+{
+    OSSL_PARAM *p;
+    int val = 0;
+
+    return TEST_ptr(p = OSSL_PARAM_locate(params, key))
+           && TEST_true(OSSL_PARAM_get_int(p, &val))
+           && TEST_int_eq(val, expected);
+}
+
+static int do_check_utf8_str(OSSL_PARAM params[], const char *key,
+                             const char *expected)
+{
+    OSSL_PARAM *p;
+    char *bufp = 0;
+    int ret;
+
+    ret = TEST_ptr(p = OSSL_PARAM_locate(params, key))
+          && TEST_true(OSSL_PARAM_get_utf8_string(p, &bufp, 0))
+          && TEST_str_eq(bufp, expected);
+    OPENSSL_free(bufp);
+    return ret;
+}
+
+static int test_dsa_todata(void)
+{
+    EVP_PKEY *pkey = NULL;
+    OSSL_PARAM *to_params = NULL, *all_params = NULL;
+    OSSL_PARAM gen_params[4];
+    int ret = 0;
+    const unsigned char *pkeydata = dsa_key;
+
+    unsigned char dsa_seed[] = {
+        0xbc, 0x8a, 0x81, 0x64, 0x9e, 0x9d, 0x63, 0xa7, 0xa3, 0x5d, 0x87, 0xdd,
+        0x32, 0xf3, 0xc1, 0x9f, 0x18, 0x22, 0xeb, 0x73, 0x63, 0xad, 0x5e, 0x7b,
+        0x90, 0xc1, 0xe3, 0xe0
+    };
+    int dsa_pcounter = 319;
+    int dsa_gindex = 5;
+
+    gen_params[0] = OSSL_PARAM_construct_octet_string(OSSL_PKEY_PARAM_FFC_SEED,
+                                                      (void*)dsa_seed,
+                                                      sizeof(dsa_seed));
+    gen_params[1] = OSSL_PARAM_construct_int(OSSL_PKEY_PARAM_FFC_GINDEX,
+                                             &dsa_gindex);
+    gen_params[2] = OSSL_PARAM_construct_int(OSSL_PKEY_PARAM_FFC_PCOUNTER,
+                                             &dsa_pcounter);
+    gen_params[3] = OSSL_PARAM_construct_end();
+
+    if (!TEST_ptr(pkey = d2i_AutoPrivateKey_ex(NULL, &pkeydata, sizeof(dsa_key),
+                                               mainctx, NULL))
+        || !TEST_int_eq(EVP_PKEY_todata(pkey, EVP_PKEY_KEYPAIR, &to_params), 1)
```

Comment:
Should there be a round trip test here? i.e. confirm that you can recreate an equivalent key by using EVP_PKEY_fromdata() on the returned params?

Reply:
ok..

Reply:
Yes - added..